mountpoint-s3-client/src/s3_crt_client/get_object.rs

```diff
@@ -67,15 +73,43 @@ impl S3CrtClient {
 
         let mut options = S3CrtClientInner::new_meta_request_options(message, S3Operation::GetObject);
         options.part_size(self.inner.read_part_size as u64);
+
+        let (object_metadata_sender, object_metadata_receiver) = oneshot::channel::<HashMap<String, String>>();
+        let object_metadata_sender_holder = Arc::new(Mutex::new(Some(object_metadata_sender)));
+        let object_metadata_sender_on_headers = object_metadata_sender_holder.clone();
+        let object_metadata_sender_on_finish = object_metadata_sender_holder.clone();
+        let lazy_object_metadata = Lazy::new(object_metadata_receiver);
+
         let request = self.inner.make_meta_request_from_options(
             options,
             span,
             |_| (),
-            |_, _| (),
+            move |headers, status| {
+                // Headers can be returned multiple times, but the object metadata doesn't change.
+                // Explicitly ignore the case where we've already set object metadata.
+                if let Some(object_metadata_sender) = object_metadata_sender_on_headers.lock().unwrap().take() {
+                    // On a 4xx / 5xx, don't send anything and cancel the future.
+                    // TODO - What happens on a 3xx status code?
```

Comment:
I'd move the status check before the `take()`. We only send the metadata on a successful request and rely on the `on_finish` callback to clean up if none succeed.